### DISCUSSION TITLE
Seeking memory guidance  (for python implementation).

### DISCUSSION THREAD

**Discussion by jaimerodriguez** [microsoft/semantic-kernel#8377] [Q&A] [answered]
I am a noob to SK. I am coding in python.   the usage of memory is unclear to me, as documentation is scarce and implementation details don't convey enough context. 

1. Is Memory (e.g. SemanticTextMemory) supposed to be for insights shared across the LLM?   
[Should i not store user-specific details there? 

2. How can I leverage memory from within my plugins?   I would like to both save stuff into memory (most important) and also to retrieve stuff from it. 

3. I am using OpenAI's model directly. Is there any implementation of memory that goes against an offering from them?  Right now, i am using Azure AI Search as the storage, but would have preferred to keep it to OpenAI storage, if possible. 

Thanks!

**Comment by AumGupta** [comment, marked as answer]
Hey @jaimerodriguez,

> 1. Is Memory (e.g. SemanticTextMemory) supposed to be for insights shared across the LLM? Should i not store user-specific details there?

**Yes**, SemanticTextMemory (STM) is primarily intended for storing general knowledge or facts that can be shared across different LLM instances. It's **not** recommended to store user-specific details in STM. This is because STM is designed to be a shared resource, and storing sensitive user data could lead to privacy concerns.

> 2. How can I leverage memory from within my plugins?   I would like to both save stuff into memory (most important) and also to retrieve stuff from it.

To leverage memory within your plugins, you can use the following methods:

* **Saving to Memory:**
  - Use the `llm.save_context()` method to store data in the LLM's memory. This data will be available for future interactions.
  - You can store any type of data, but it's generally best to store structured data (e.g., dictionaries, lists) for easier retrieval.

* **Retrieving from Memory:**
  - Use the `llm.load_context()` method to retrieve data from the LLM's memory.
  - You can specify the specific context you want to retrieve, or retrieve the entire context.

> 3. I am using OpenAI's model directly. Is there any implementation of memory that goes against an offering from them?  Right now, i am using Azure AI Search as the storage, but would have preferred to keep it to OpenAI storage, if possible.

**OpenAI doesn't provide a built-in memory system for storing user-specific data.** While you can use the `llm.save_context()` method to store data within the LLM's memory, this data is not persistent and may be lost if the LLM instance is restarted.

If you need to store persistent user-specific data, you'll need to use an external storage solution. Azure AI Search is a good option for this, as it provides a scalable and flexible way to store and retrieve data.

> **Note:** While OpenAI doesn't offer a direct memory storage solution, you could potentially explore using a third-party service that integrates with OpenAI's APIs to provide memory functionality. However, this would require additional research and configuration.

Hope this answers your queries.
